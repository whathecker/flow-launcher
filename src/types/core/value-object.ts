export type PriorityTier = "n/a" | "highest" | "high" | "mid" | "low";

export type PriorityMeasure = "n/a" | "yes" | "no"; // TODO: consider to use MIXED type (boolean and null) instead?

export class Priority {
  constructor(
    public tier: PriorityTier,
    public importance: PriorityMeasure,
    public urgency: PriorityMeasure,
  ) {}
}
