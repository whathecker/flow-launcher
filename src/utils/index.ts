export { default as navigationRef } from "./navigationRef";
export { default as taskFilters } from "./taskFilters";
export { default as colorRenderer } from "./colorRenderer";
export { default as labelRenderer } from "./labelRenderer";
