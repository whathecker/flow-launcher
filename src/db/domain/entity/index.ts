import { Priority } from "../value-object";

export type ID = number | string;

export type Reminder = "daily" | "three_days" | "seven_days";

export class Goal {
  constructor(
    public id: ID,
    public motivation: string,
    public reminder: Reminder,
    public tasks: Task[],
  ) {}
}

export type TaskStatus = "open" | "finished";

export class Task {
  constructor(
    public title: string,
    public description: string,
    public status: TaskStatus,
    public priority?: Priority,
    public id?: ID,
  ) {}
}
