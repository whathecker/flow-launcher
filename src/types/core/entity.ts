import { Priority } from "./value-object";

export type ID = number | string;

export type GoalStatus = "open" | "finished";
export type Reminder = "daily" | "three_days" | "seven_days";

export class Goal {
  constructor(
    public status: GoalStatus,
    public title: string,
    public motivation: string,
    public reminder: Reminder,
    public tasks: Task[],
    public _id?: ID,
  ) {}
}

export interface IAddGoalInput {
  title: string;
  motivation: string;
  reminder: string;
}

export interface IAddTaskInput {
  title: string;
  description: string;
  goal_id: string;
}

export interface IFetchTasksInput {
  goal_id: string;
}

export type TaskStatus = "open" | "finished";

export class Task {
  constructor(
    public _id: ID,
    public goal_id: ID,
    public title: string,
    public status: TaskStatus,
    public description?: string,
    public priority?: Priority,
  ) {}
}
