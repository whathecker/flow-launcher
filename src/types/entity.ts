import { Priority } from "./value-object";

export type ID = number | string;

export type GoalStatus = "open" | "finished";
export type Reminder = "daily" | "three_days" | "seven_days";

export class Goal {
  constructor(
    public status: GoalStatus,
    public title: string,
    public motivation: string,
    public reminder: Reminder,
    public tasks: Task[] | [],
    public _id?: ID,
  ) {}
}

export interface IAddGoalInput {
  title: string;
  motivation: string;
  reminder: string;
}

export type TaskStatus = "open" | "finished";

export class Task {
  constructor(
    public _id: ID,
    public goal_id: ID,
    public title: string,
    public description: string,
    public status: TaskStatus,
    public priority?: Priority,
  ) {}
}
