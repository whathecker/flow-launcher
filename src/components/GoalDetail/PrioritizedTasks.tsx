import React from "react";
import { StyleSheet } from "react-native";
import { EmptyPrioritizedTasks, TaskBucketByPrio } from "./components";
import { View, Text } from "../Themed";
import { Color, Shadow, Typography } from "../../styles";
import { Task, GoalColor } from "../../types/core/entity";

type PrioritizedTasksProps = {
  highest: Task[];
  high: Task[];
  mid: Task[];
  low: Task[];
  goalColor: GoalColor;
};

function _isTasksEmpty(input: Task[]): boolean {
  return input.length === 0;
}

const PrioritizedTasks: React.FC<PrioritizedTasksProps> = ({
  highest,
  high,
  mid,
  low,
  goalColor,
}: PrioritizedTasksProps) => {
  const mergedTasks = highest.concat(high).concat(mid).concat(low);
  const isTasksEmptty = _isTasksEmpty(mergedTasks);

  return (
    <View style={styles.wrapper}>
      <View style={styles.headerAreaWrapper}>
        <Text style={styles.headerText}>{"Prioritized Tasks"}</Text>
      </View>
      {isTasksEmptty ? (
        <EmptyPrioritizedTasks />
      ) : (
        <View>
          <TaskBucketByPrio
            prio={"highest"}
            tasks={highest}
            goalColor={goalColor}
          />
          <TaskBucketByPrio prio={"high"} tasks={high} goalColor={goalColor} />
          <TaskBucketByPrio prio={"mid"} tasks={mid} goalColor={goalColor} />
          <TaskBucketByPrio prio={"low"} tasks={low} goalColor={goalColor} />
        </View>
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  wrapper: {
    width: "90%",
    borderWidth: 0.5,
    borderColor: Color.light.defaultBorder,
    borderRadius: 5,
    ...Shadow.regularbackDrop,
    padding: 20,
  },
  headerAreaWrapper: {
    marginBottom: 10,
  },
  headerText: {
    ...Typography.h4,
    fontSize: 18,
  },
});

export default PrioritizedTasks;
