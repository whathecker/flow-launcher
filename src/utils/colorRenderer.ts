import { GoalColor } from "../types/core/entity";
import {
  PriorityTier,
  PriorityBucketColor,
  ColorShadeForFirstGoalColor,
  ColorShadeForSecondGoalColor,
  ColorShadeForThirdGoalColor,
  ColorShadeForFourthGoalColor,
} from "../types/core/value-object";

const assignColorForGoal = (index: number): GoalColor => {
  if (_isFirstColor(index)) return "#E9695B";

  let color;
  switch (_findColorToRender(index)) {
    case 1:
      color = "#84C5E8";
      break;
    case 2:
      color = "#1C88BA";
      break;
    case 3:
      color = "#F0B541";
      break;
    default:
      color = "#E9695B";
      break;
  }

  return color as GoalColor;
};

function _isFirstColor(index: number): boolean {
  if (index === 0 || index % 4 === 0) return true;

  return false;
}

function _findColorToRender(index: number): number {
  return index % 4;
}

const getColorForPrioBucket = (
  goalColor: GoalColor,
  prio: PriorityTier,
): PriorityBucketColor | "#FEFEF8" => {
  let color;
  switch (goalColor) {
    case "#E9695B":
      color = _findColorShadeForFirstColor(prio);
      break;
    case "#84C5E8":
      color = _findColorShadeForSecondColor(prio);
      break;
    case "#1C88BA":
      color = _findColorShadeForThirdColor(prio);
      break;
    case "#F0B541":
      color = _findColorShadeForFourthColor(prio);
      break;
    default:
      color = "#FEFEF8";
      break;
  }
  return color as PriorityBucketColor;
};

function _findColorShadeForFirstColor(
  prio: PriorityTier,
): ColorShadeForFirstGoalColor | "#FEFEF8" {
  let color;
  switch (prio) {
    case "highest":
      color = "#C2571A";
      break;
    case "high":
      color = "#DA621E";
      break;
    case "mid":
      color = "#F26D21";
      break;
    case "low":
      color = "#F58B4C";
      break;
    default:
      color = "#FEFEF8";
      break;
  }
  return color as ColorShadeForFirstGoalColor;
}

function _findColorShadeForSecondColor(
  prio: PriorityTier,
): ColorShadeForSecondGoalColor | "#FEFEF8" {
  let color;
  switch (prio) {
    case "highest":
      color = "#107896";
      break;
    case "high":
      color = "#0D9BC3";
      break;
    case "mid":
      color = "#2FAFD3";
      break;
    case "low":
      color = "#65CBE9";
      break;
    default:
      color = "#FEFEF8";
      break;
  }
  return color as ColorShadeForSecondGoalColor;
}

function _findColorShadeForThirdColor(
  prio: PriorityTier,
): ColorShadeForThirdGoalColor | "#FEFEF8" {
  let color;
  switch (prio) {
    case "highest":
      color = "#093145";
      break;
    case "high":
      color = "#0D3D56";
      break;
    case "mid":
      color = "#0D5275";
      break;
    case "low":
      color = "#3C6478";
      break;
    default:
      color = "#FEFEF8";
      break;
  }
  return color as ColorShadeForThirdGoalColor;
}

function _findColorShadeForFourthColor(
  prio: PriorityTier,
): ColorShadeForFourthGoalColor | "#FEFEF8" {
  let color;
  switch (prio) {
    case "highest":
      color = "#BCA136";
      break;
    case "high":
      color = "#D3B53D";
      break;
    case "mid":
      color = "#EBC944";
      break;
    case "low":
      color = "#EFD469";
      break;
    default:
      color = "#FEFEF8";
      break;
  }
  return color as ColorShadeForFourthGoalColor;
}

export default { assignColorForGoal, getColorForPrioBucket };
