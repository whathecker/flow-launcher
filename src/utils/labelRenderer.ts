import { PriorityTier } from "../types/core/value-object";

const renderPrioBucketLabel = (prio: PriorityTier): string => {
  let label = "";
  switch (prio) {
    case "highest":
      label = "Highest priority";
      break;
    case "high":
      label = "High priority";
      break;
    case "mid":
      label = "Mid priority";
      break;
    case "low":
      label = "Low priority";
      break;
    default:
      // TODO: Add error handling here
      break;
  }
  return label;
};

export default { renderPrioBucketLabel };
