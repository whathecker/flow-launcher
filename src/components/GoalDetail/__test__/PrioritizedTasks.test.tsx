import * as React from "react";
import renderer from "react-test-renderer";
import PrioritizedTasks from "../PrioritizedTasks";
import { Task } from "../../../types/core/entity";
import { taskFilters } from "../../../utils";

describe("Test PrioritizedTasks component", () => {
  it("Should renders PrioritizedTasks correctly", () => {
    const tasks: Task[] = [
      {
        _id: "622605f8766c2b210fcc5890",
        description: "",
        goal_id: "61d024a2b4cda2847a224bdf",
        priority: {
          importance: "yes",
          tier: "highest",
          urgency: "yes",
        },
        status: "open",
        title: "Test1",
      },
      {
        _id: "622605fd766c2b210fcc5891",
        description: "",
        goal_id: "61d024a2b4cda2847a224bdf",
        priority: {
          importance: "yes",
          tier: "highest",
          urgency: "yes",
        },
        status: "open",
        title: "Test2",
      },
      {
        _id: "62260602766c2b210fcc5892",
        description: "",
        goal_id: "61d024a2b4cda2847a224bdf",
        priority: {
          importance: "yes",
          tier: "high",
          urgency: "no",
        },
        status: "open",
        title: "test3",
      },
      {
        _id: "62260606766c2b210fcc5893",
        description: "",
        goal_id: "61d024a2b4cda2847a224bdf",
        priority: {
          importance: "yes",
          tier: "high",
          urgency: "no",
        },
        status: "open",
        title: "Test4",
      },
      {
        _id: "62260609766c2b210fcc5894",
        description: "",
        goal_id: "61d024a2b4cda2847a224bdf",
        priority: {
          importance: "no",
          tier: "low",
          urgency: "no",
        },
        status: "open",
        title: "test5",
      },
    ];
    const result = taskFilters.filterByPriorityScheme(tasks);
    const tree = renderer
      .create(
        <PrioritizedTasks
          highest={result.highest}
          high={result.high}
          mid={result.mid}
          low={result.low}
        />,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
