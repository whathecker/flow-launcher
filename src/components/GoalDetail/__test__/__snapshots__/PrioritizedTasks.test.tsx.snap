// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test PrioritizedTasks component Should renders PrioritizedTasks correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
      },
      Object {
        "borderColor": "#636262",
        "borderRadius": 5,
        "borderWidth": 0.5,
        "padding": 20,
        "shadowColor": "#171717",
        "shadowOffset": Object {
          "height": 5,
          "width": 3,
        },
        "shadowOpacity": 0.2,
        "shadowRadius": 3,
        "width": "85%",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
        },
        Object {
          "marginBottom": 10,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000000",
          },
          Object {
            "fontFamily": "EB-Garamond",
            "fontSize": 18,
          },
        ]
      }
    >
      Prioritized Tasks
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
        },
        Object {
          "alignItems": "center",
          "flexDirection": "column",
          "height": "80%",
          "justifyContent": "center",
          "padding": 20,
        },
      ]
    }
  >
    <Image
      source={1}
      style={
        Object {
          "height": 40,
          "width": 40,
        }
      }
    />
    <Text
      style={
        Array [
          Object {
            "color": "#000000",
          },
          Object {
            "color": "#848484",
            "fontFamily": "Forum-Regular",
            "fontSize": 16,
            "paddingTop": 20,
          },
        ]
      }
    >
      Small steps, great achievement
    </Text>
  </View>
</View>
`;
